src/plugins_types/date_and_time.c

```diff
@@ -0,0 +1,274 @@
+/**
+ * @file date_and_time.c
+ * @author Michal Vasko <mvasko@cesnet.cz>
+ * @brief ietf-yang-types date-and-time type plugin.
+ *
+ * Copyright (c) 2019-2021 CESNET, z.s.p.o.
+ *
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#define _GNU_SOURCE
+
+#include "plugins_types.h"
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "libyang.h"
+
+#include "compat.h"
+#include "common.h"
+
+/**
+ * @brief Convert string to a number.
+ *
+ * @param[in,out] str String to convert, the parsed number are skipped.
+ * @param[in] len Expected length of the number. 0 to parse at least 1 character.
+ * @param[in] full_str Full string for error generation.
+ * @param[out] num Converted number.
+ * @param[out] err Error information on error.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+convert_number(const char **str, int len, const char *full_str, int *num, struct ly_err_item **err)
+{
+    char *ptr;
+
+    *num = strtoul(*str, &ptr, 10);
+    if ((len && (ptr - *str != len)) || (!len && (ptr == *str))) {
+        return ly_err_new(err, LY_EVALID, LYVE_DATA, NULL, NULL, "Invalid character '%c' in date-and-time value \"%s\", "
+                "a digit expected.", ptr[0], full_str);
+    }
+
+    *str = ptr;
+    return LY_SUCCESS;
+}
+
+/**
+ * @brief Check a single character.
+ *
+ * @param[in,out] str String to check, the parsed character is skipped.
+ * @param[in] c Array of possible characters.
+ * @param[in] full_str Full string for error generation.
+ * @param[out] err Error information on error.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+check_char(const char **str, char c[], const char *full_str, struct ly_err_item **err)
+{
+    LY_ERR ret;
+    uint32_t i;
+    char *exp_str;
+
+    for (i = 0; c[i]; ++i) {
+        if ((*str)[0] == c[i]) {
+            break;
+        }
+    }
+    if (!c[i]) {
+        exp_str = malloc(3 + (i - 1) * 7 + 1);
```

Comment:
some comment explaining these magic numbers would be useful here

Reply:
Okay, I have tried.